src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -140,18 +146,45 @@ private static void checkpointRestore1() throws
             }
         }
 
+        RestoreException restoreException = null;
         try {
             globalContext.afterRestore(null);
         } catch (RestoreException re) {
             if (checkpointException == null) {
-                throw re;
+                restoreException = re;
+            } else {
+                for (Throwable t : re.getSuppressed()) {
+                    checkpointException.addSuppressed(t);
+                }
             }
-            for (Throwable t : re.getSuppressed()) {
-                checkpointException.addSuppressed(t);
+        }
+
+        if (newArguments != null && newArguments.length() > 0) {
+            String[] args = newArguments.split(" ");
+            if (args.length > 0) {
+                try {
+                    Class<?> newMainClass = Class.forName(args[0], false,
+                        callerClass.getClassLoader());
+                    Method newMain = newMainClass.getDeclaredMethod("main",
+                        String[].class);
+                    newMain.invoke(null,
+                        (Object)Arrays.copyOfRange(args, 1, args.length));
+                } catch (Throwable e) {
```

Comment:
Is `Throwable` the right thing to catch here?  I think Error-subclasses, like OutOfMemoryError or VerifyError, should propagate past here without being added to the suppressed set of the `restoreException`.